README.md
# Veritaserum

[![Build Status](https://travis-ci.org/uesteibar/veritaserum.svg?branch=master)](https://travis-ci.org/uesteibar/veritaserum)
[![Hex Version](https://img.shields.io/hexpm/v/veritaserum.svg)](https://hex.pm/packages/veritaserum)

Simple sentiment analysis for [Elixir](http://elixir-lang.org/) based on the AFINN-165 list and some extra enhancements.

It also supports:
- **emojis** (❤️, 😱...)
- **boosters** (*very*, *really*...)
- **negators** (*don't*, *not*...).

## Index

- [Installation](#installation)
- [Usage](#usage)
- [Running Locally](#running-locally)
- [Contributing](#contributing)

## Installation

Add `veritaserum` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [{:veritaserum, "~> 0.2.2"}]
end
```

## Usage

To analyze a text
```elixir
Veritaserum.analyze("I love Veritaserum!") #=> 3

# It also supports emojis
Veritaserum.analyze("I ❤️ Veritaserum!") #=> 2

# It also supports negators
Veritaserum.analyze("I like Veritaserum!") #=> 2
Veritaserum.analyze("I don't like Veritaserum!") #=> -2

#and boosters
Veritaserum.analyze("Veritaserum is cool!") #=> 1
Veritaserum.analyze("Veritaserum is very cool!") #=> 2
```

You can also pass a list
```elixir
Veritaserum.analyze(["I love Veritaserum!", "I hate some things!"]) #=> [3, -3]
```

Documentation can be found on [HexDocs](https://hexdocs.pm/veritaserum).

## Running locally

Clone the repository
```bash
git clone git@github.com:uesteibar/veritaserum.git
```

Install dependencies
```bash
cd veritaserum
mix deps.get
```

To run the tests
```bash
mix test
```

To run the lint
```elixir
mix credo
```

## Contributing

Pull requests are always welcome =)

The project uses [standard-changelog](https://github.com/conventional-changelog/conventional-changelog) to update the [Changelog](https://github.com/uesteibar/veritaserum/blob/master/CHANGELOG.md) with each commit message and upgrade the package version.
For that reason every contribution should have a title and body that follows the [conventional commits standard](https://conventionalcommits.org/) conventions (e.g. `feat(analyzer): Make it smarter than Jarvis`).

To make this process easier, you can do the following:

Install `commitizen` and `cz-conventional-changelog` globally
```bash
npm i -g commitizen cz-conventional-changelog
```

Save `cz-conventional-changelog` as default
```bash
echo '{ "path": "cz-conventional-changelog" }' > ~/.czrc
```

Instead of `git commit`, you can now run
```
git cz
```
and follow the instructions to generate the commit message.


CHANGELOG.md
<a name="0.2.2"></a>
# [0.2.2](https://github.com/uesteibar/veritaserum/compare/v0.2.1...v) (2020-02-11)


### Bug Fixes

* **deprecations:** Fix deprecation warnings ([0d2a6b3](https://github.com/uesteibar/veritaserum/commit/0d2a6b3))



<a name="0.2.1"></a>
## [0.2.1](https://github.com/uesteibar/veritaserum/compare/v0.2.0...v0.2.1) (2019-04-28)


### Features

* **analyze:** change the way we split the string on emoticons to enhance perfomance


<a name="0.2.0"></a>
## [0.2.0](https://github.com/uesteibar/veritaserum/compare/v0.1.3...v0.2.0) (2017-09-19)


### Features

* **analyze:** Add option return: :score_and_marks to inspect result in detail ([a0ead8e](https://github.com/uesteibar/veritaserum/commit/a0ead8e))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/uesteibar/veritaserum/compare/v0.1.2...v0.1.3) (2017-09-07)


### Bug Fixes

* **evaluator:** Evaluate emojis without surrounding spaces ([8af2690](https://github.com/uesteibar/veritaserum/commit/8af2690))


<a name="0.1.2"></a>
## [0.1.2](https://github.com/uesteibar/veritaserum/compare/v0.1.1...v0.1.2) (2017-05-17)


### Bug Fixes

* **elixir_version:** Set minimum elixir version to 1.3 ([5747dd6](https://github.com/uesteibar/veritaserum/commit/5747dd6))


<a name="0.1.1"></a>
## [0.1.1](https://github.com/uesteibar/veritaserum/compare/v0.1.0...v0.1.1) (2017-05-17)


### Performance Improvements

* **evaluator:** Read word files and generate methods on compilation ([d34bfcd](https://github.com/uesteibar/veritaserum/commit/d34bfcd))


<a name="0.1.0"></a>
# 1.0.0 (2017-05-17)


test/veritaserum_test.exs
defmodule VeritaserumTest do
  use ExUnit.Case

  doctest Veritaserum

  import Veritaserum

  describe "when given a list" do
    @list ["I love Veritaserum", "I hate nothing"]

    test "calculates sentimental value for each" do
      assert analyze(@list) == [3, -3]
    end
  end

  describe "when text has no relevant word" do
    @text "I build Veritaserum"

    test "sentimental value is 0" do
      assert analyze(@text) == 0
    end
  end

  describe "when text has relevant word" do
    @text "I love Veritaserum"

    test "calculates sentimental value" do
      assert analyze(@text) == 3
    end
  end

  describe "when return option is :score_and_marks" do
    @text "You really love Veritaserum. Don't you?"

    test "returns score with marks" do
      result =
        {4,
         [
           {:neutral, 0, "you"},
           {:booster, 1, "really"},
           {:word, 3, "love"},
           {:neutral, 0, "veritaserum"},
           {:negator, 1, "don't"},
           {:neutral, 0, "you?"}
         ]}

      assert ^result = analyze(@text, return: :score_and_marks)
    end
  end

  describe "when text has relevant emoji" do
    @text "I ❤️ Veritaserum"

    test "calculates sentimental value" do
      assert analyze(@text) == 3
    end

    @text "I❤️Veritaserum"

    test "calculates sentimental value of emoji without spaces" do
      assert analyze(@text) == 3
    end
  end

  describe "when text has irrelevant characters" do
    @text "I love! Veritaserum"

    test "removes irrelevant characters" do
      assert analyze(@text) == 3
    end
  end

  describe "when text has negator" do
    @text "I don't hate Veritaserum"

    test "the negator flips the value of the next word" do
      assert analyze(@text) == 3
    end
  end

  describe "when positive word has booster" do
    @text "I really love Veritaserum"

    test "the booster increases the value of the next word" do
      assert analyze(@text) == 4
    end
  end

  describe "when negative word has booster" do
    @text "I really hate Veritaserum"

    test "the booster decreases the value of the next word" do
      assert analyze(@text) == -4
    end
  end

  describe "when irrelevant word has booster" do
    @text "I really buy Veritaserum"

    test "the booster does not affect the value of the next word" do
      assert analyze(@text) == 0
    end
  end
end


.credo.exs
%{
  configs: [
    %{
      name: "default",
      files: %{
        included: ["lib/"],
        excluded: [~r"/_build/", ~r"/deps/"]
      },
      check_for_updates: true,
      strict: true,
      color: true,
      checks: [
        {Credo.Check.Consistency.ExceptionNames},
        {Credo.Check.Consistency.LineEndings},
        {Credo.Check.Consistency.MultiAliasImportRequireUse},
        {Credo.Check.Consistency.ParameterPatternMatching},
        {Credo.Check.Consistency.SpaceAroundOperators},
        {Credo.Check.Consistency.SpaceInParentheses},
        {Credo.Check.Consistency.TabsOrSpaces},
        {Credo.Check.Design.AliasUsage, priority: :low},
        {Credo.Check.Design.DuplicatedCode, excluded_macros: []},
        {Credo.Check.Design.TagTODO, exit_status: 2},
        {Credo.Check.Design.TagFIXME},
        {Credo.Check.Readability.FunctionNames},
        {Credo.Check.Readability.LargeNumbers},
        {Credo.Check.Readability.MaxLineLength, priority: :low, max_length: 80},
        {Credo.Check.Readability.ModuleAttributeNames},
        {Credo.Check.Readability.ModuleDoc},
        {Credo.Check.Readability.ModuleNames},
        {Credo.Check.Readability.ParenthesesOnZeroArityDefs},
        {Credo.Check.Readability.ParenthesesInCondition},
        {Credo.Check.Readability.PredicateFunctionNames},
        {Credo.Check.Readability.PreferImplicitTry},
        {Credo.Check.Readability.RedundantBlankLines},
        {Credo.Check.Readability.StringSigils},
        {Credo.Check.Readability.TrailingBlankLine},
        {Credo.Check.Readability.TrailingWhiteSpace},
        {Credo.Check.Readability.VariableNames},
        {Credo.Check.Readability.Semicolons},
        {Credo.Check.Readability.SpaceAfterCommas},
        {Credo.Check.Refactor.DoubleBooleanNegation},
        {Credo.Check.Refactor.CondStatements},
        {Credo.Check.Refactor.CyclomaticComplexity},
        {Credo.Check.Refactor.FunctionArity},
        {Credo.Check.Refactor.MatchInCondition},
        {Credo.Check.Refactor.NegatedConditionsInUnless},
        {Credo.Check.Refactor.NegatedConditionsWithElse},
        {Credo.Check.Refactor.Nesting},
        {Credo.Check.Refactor.PipeChainStart, false},
        {Credo.Check.Refactor.UnlessWithElse},
        {Credo.Check.Warning.BoolOperationOnSameValues},
        {Credo.Check.Warning.IExPry},
        {Credo.Check.Warning.IoInspect},
        {Credo.Check.Warning.LazyLogging},
        {Credo.Check.Warning.OperationOnSameValues},
        {Credo.Check.Warning.OperationWithConstantResult},
        {Credo.Check.Warning.UnusedEnumOperation},
        {Credo.Check.Warning.UnusedFileOperation},
        {Credo.Check.Warning.UnusedKeywordOperation},
        {Credo.Check.Warning.UnusedListOperation},
        {Credo.Check.Warning.UnusedPathOperation},
        {Credo.Check.Warning.UnusedRegexOperation},
        {Credo.Check.Warning.UnusedStringOperation},
        {Credo.Check.Warning.UnusedTupleOperation},

        # Controversial and experimental checks (opt-in, just remove `, false`)
        #
        {Credo.Check.Refactor.ABCSize, false},
        {Credo.Check.Refactor.AppendSingleItem, false},
        {Credo.Check.Refactor.VariableRebinding, false},
        {Credo.Check.Warning.MapGetUnsafePass, false},

        # Deprecated checks (these will be deleted after a grace period)
        {Credo.Check.Readability.Specs, false},
        {Credo.Check.Warning.NameRedeclarationByAssignment, false},
        {Credo.Check.Warning.NameRedeclarationByCase, false},
        {Credo.Check.Warning.NameRedeclarationByDef, false},
        {Credo.Check.Warning.NameRedeclarationByFn, false},

        # Custom checks can be created using `mix credo.gen.check`.
      ]
    }
  ]
}


lib/evaluator.ex
defmodule Veritaserum.Evaluator do
  @moduledoc """
  Evaluats words, boosters, negators and emoticons.
  """

  ["word", "emoticon", "negator", "booster"]
  |> Enum.each(fn facet ->
    File.read!("#{__DIR__}/../config/facets/#{facet}.json")
    |> Jason.decode!()
    |> (fn list ->
          def unquote(:"#{facet}_list")(),
            do: unquote(list |> Enum.map(fn {key, _} -> key end) |> List.flatten())

          list
        end).()
    |> Enum.each(fn {word, value} ->
      @doc """
      Evaluates if a word/emoji is a **#{facet}** and returns value.

          iex> Veritaserum.Evaluator.evaluate_#{facet}("#{word}")
          #{value}
      """
      def unquote(:"evaluate_#{facet}")(unquote(word)), do: unquote(value)
    end)

    def unquote(:"evaluate_#{facet}")(_), do: nil
  end)
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# 3rd-party users, it should be done in your "mix.exs" file.

# You can configure for your application as:
#
#     config :veritaserum, key: :value
#
# And access this configuration in your application as:
#
#     Application.get_env(:veritaserum, :key)
#
# Or configure a 3rd-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#
#     import_config "#{Mix.env}.exs"


lib/veritaserum.ex
defmodule Veritaserum do
  @moduledoc """
  Sentiment analisis based on AFINN-165, emojis and some enhancements.

  Also supports:
  - emojis (❤️, 😱...)
  - boosters (*very*, *really*...)
  - negators (*don't*, *not*...).
  """

  alias Veritaserum.Evaluator

  @spec analyze(list(String.t()) | String.t()) :: list(integer) | integer
  def analyze(input) when is_list(input) do
    input
    |> Stream.map(&analyze/1)
    |> Enum.to_list()
  end

  def analyze(input) do
    {score, _} = analyze(input, return: :score_and_marks)

    score
  end

  @doc """
  Returns a sentiment value for the given text

      iex> Veritaserum.analyze(["I ❤️ Veritaserum", "Veritaserum is really awesome"])
      [3, 5]

      iex> Veritaserum.analyze("I love Veritaserum")
      3
  """
  @spec analyze(String.t(), return: :score_and_marks) :: {number(), [{atom, number, String.t()}]}
  def analyze(input, return: :score_and_marks) do
    list_with_marks = get_list_with_marks(input)
    score = get_score(list_with_marks)

    {score, list_with_marks}
  end

  defp get_score(words) do
    words
    |> analyze_list
    |> Enum.sum()
  end

  defp get_list_with_marks(input) do
    input
    |> clean
    |> String.split()
    |> mark_list
    |> Enum.reverse()
  end

  defp mark_word(word) do
    with {_, nil, _} <- {:negator, Evaluator.evaluate_negator(word), word},
         {_, nil, _} <- {:booster, Evaluator.evaluate_booster(word), word},
         {_, nil, _} <- {:emoticon, Evaluator.evaluate_emoticon(word), word},
         {_, nil, _} <- {:word, Evaluator.evaluate_word(word), word},
         do: {:neutral, 0, word}
  end

  defp mark_list([head | tail]) do
    mark_list(tail, [mark_word(head)])
  end

  defp mark_list([head | tail], result) do
    mark_list(tail, [mark_word(head) | result])
  end

  defp mark_list([], result), do: result

  defp analyze_mark({type, score, _}) do
    case type do
      :word -> score
      :emoticon -> score
      _ -> 0
    end
  end

  defp analyze_mark(mark, previous) do
    case previous do
      {:negator, _, _} ->
        -analyze_mark(mark)

      {:booster, booster_value, _} ->
        analyze_mark(mark) |> apply_booster(booster_value)

      _ ->
        analyze_mark(mark)
    end
  end

  defp analyze_list([head | tail]) do
    analyze_list(tail, head, [analyze_mark(head)])
  end

  defp analyze_list([head | tail], previous, result) do
    analyze_list(tail, head, [analyze_mark(head, previous) | result])
  end

  defp analyze_list([], _, result), do: result

  defp apply_booster(word_value, booster) when word_value > 0, do: word_value + booster
  defp apply_booster(word_value, booster) when word_value < 0, do: word_value - booster
  defp apply_booster(word_value, _booster), do: word_value

  defp clean(text) do
    text
    |> String.replace(~r/\n/, " ")
    |> String.downcase()
    |> String.replace(~r/[.,\/#!$%\^&\*;:{}=_`\"~()]/, " ")
    |> :binary.replace(Evaluator.emoticon_list(), "  ", insert_replaced: 1)
    |> String.replace(~r/ {2,}/, " ")
  end
end


mix.exs
defmodule Veritaserum.Mixfile do
  use Mix.Project

  @version "0.2.2"

  def project do
    [
      app: :veritaserum,
      version: @version,
      elixir: "~> 1.3",
      description: "Sentiment analysis based on afinn-165, emojis and some enhancements.",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      test_coverage: [tool: Coverex.Task],
      aliases: aliases(),
      package: package(),
      deps: deps(),
      docs: docs(),
      source_url: github()
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp package do
    [
      files: [
        "config/facets/word.json",
        "config/facets/negator.json",
        "config/facets/booster.json",
        "config/facets/emoticon.json",
        "lib",
        "mix.exs",
        "README.md",
        "LICENSE.md"
      ],
      links: %{"github" => github()},
      maintainers: ["Unai Esteibar <uesteibar@gmail.com>"],
      licenses: ["Apache 2.0"]
    ]
  end

  defp docs do
    [
      source_ref: "v#{@version}",
      main: "Veritaserum",
      logo: "veritaserum_logo.png",
      extras: ["README.md"]
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.0.0", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.19", only: :dev, runtime: false},
      {:dialyxir, "~> 0.5", only: [:dev], runtime: false},
      {:coverex, "~> 1.4", only: :test}
    ]
  end

  defp aliases do
    [
      test: "test --cover"
    ]
  end

  defp github do
    "https://github.com/uesteibar/veritaserum"
  end
end
